File: benchmarks/fixtures/webapp_swift/Sources/webapp/validators/Validators.swift
import Foundation

/// Conformance for any type that can validate itself.
protocol Validating {
    func validate() throws
}

/// Validates user-supplied registration data.
struct UserValidator: Validating {
    let email: String

    func validate() throws {
        if !email.contains("@") {
            throw ValidationError(field: "email", message: "invalid email")
        }
    }
}

/// Validates payment payloads.
struct PaymentValidator: Validating {
    let amount: Int

    func validate() throws {
        if amount <= 0 {
            throw ValidationError(field: "amount", message: "must be positive")
        }
    }
}

File: benchmarks/fixtures/webapp_kt/src/main/kotlin/webapp/validators/Validators.kt
package webapp.validators

import webapp.errors.ValidationError

/** Conformance for any type that can validate itself. */
interface Validating {
    fun validate()
}

/** Validates user-supplied registration data. */
class UserValidator(val email: String) : Validating {
    override fun validate() {
        if (!email.contains("@")) {
            throw ValidationError("email", "invalid email")
        }
    }
}

/** Validates payment payloads. */
class PaymentValidator(val amount: Int) : Validating {
    override fun validate() {
        if (amount <= 0) {
            throw ValidationError("amount", "must be positive")
        }
    }
}
